File: programmers/lv2/연속된 부분 수열의 합/연속된 부분 수열의 합.js
// https://school.programmers.co.kr/learn/courses/30/lessons/178870

const sequence = [2, 2, 2, 2, 2];
const k = 6;

const solution = (sequence, k) => {
  let result = [];
  let sum = 0;
  let lidx = 0;
  let ridx = -1;

  while (true) {
    if (sum < k) {
      ridx += 1;
      if (ridx >= sequence.length) {
        break;
      }
      sum += sequence[ridx];
    } else {
      sum -= sequence[lidx];
      if (lidx >= sequence.length) {
        break;
      }
      lidx += 1;
    }

    if (sum === k) {
      result.push([lidx, ridx]);
    }
  }

  const arr = result.sort(([x1, y1], [x2, y2]) => {
    const sub1 = y1 - x1;
    const sub2 = y2 - x2;

    if (sub1 !== sub2) {
      return sub1 - sub2;
    } else {
      return x1 - x2;
    }
  });

  return arr[0];
};
console.log(solution(sequence, k));

File: programmers/lv2/연속된 부분 수열의 합/연속된 부분 수열의 합.py
# https://school.programmers.co.kr/learn/courses/30/lessons/178870
    
def solution(sequence, k):
  result = []
  sum = 0
  lidx = 0
  ridx = -1
    
  while True:
    if sum < k:
      ridx += 1
      if ridx >= len(sequence):
        break

      sum += sequence[ridx]
    else:
      sum -= sequence[lidx]
      if lidx >= len(sequence):
        break
      lidx += 1
    if sum == k:
      result.append([lidx, ridx])
    
  result.sort(key=lambda x: (x[1]-x[0], x[0]))
  return result[0]



sequence = [2, 2, 2, 2, 2]
k = 6
print(solution(sequence, k))